File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/journalpostkobling/JournalpostSedKoblingService.java
package no.nav.melosys.eessi.service.journalpostkobling;

import java.util.Optional;

import io.getunleash.Unleash;
import lombok.extern.slf4j.Slf4j;
import no.nav.melosys.eessi.config.featuretoggle.ToggleName;
import no.nav.melosys.eessi.integration.eux.case_store.CaseStoreConsumer;
import no.nav.melosys.eessi.integration.eux.case_store.CaseStoreDto;
import no.nav.melosys.eessi.integration.saf.SafConsumer;
import no.nav.melosys.eessi.kafka.producers.model.MelosysEessiMelding;
import no.nav.melosys.eessi.models.FagsakRinasakKobling;
import no.nav.melosys.eessi.models.JournalpostSedKobling;
import no.nav.melosys.eessi.models.SedType;
import no.nav.melosys.eessi.models.buc.BUC;
import no.nav.melosys.eessi.models.buc.Document;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.repository.JournalpostSedKoblingRepository;
import no.nav.melosys.eessi.service.eux.EuxService;
import no.nav.melosys.eessi.service.saksrelasjon.SaksrelasjonService;
import no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding.MelosysEessiMeldingMapperFactory;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class JournalpostSedKoblingService {

    private final JournalpostSedKoblingRepository journalpostSedKoblingRepository;
    private final CaseStoreConsumer caseStoreConsumer;
    private final EuxService euxService;
    private final SaksrelasjonService saksrelasjonService;
    private final SafConsumer safConsumer;
    private final MelosysEessiMeldingMapperFactory melosysEessiMeldingMapperFactory;
    private final Unleash unleash;

    public JournalpostSedKoblingService(
            JournalpostSedKoblingRepository journalpostSedKoblingRepository,
            CaseStoreConsumer caseStoreConsumer,
            EuxService euxService,
            SaksrelasjonService saksrelasjonService,
            SafConsumer safConsumer,
            MelosysEessiMeldingMapperFactory melosysEessiMeldingMapperFactory, Unleash unleash) {
        this.journalpostSedKoblingRepository = journalpostSedKoblingRepository;
        this.caseStoreConsumer = caseStoreConsumer;
        this.euxService = euxService;
        this.saksrelasjonService = saksrelasjonService;
        this.safConsumer = safConsumer;
        this.melosysEessiMeldingMapperFactory = melosysEessiMeldingMapperFactory;
        this.unleash = unleash;
    }

    public Optional<JournalpostSedKobling> finnVedJournalpostID(String journalpostID) {
        return journalpostSedKoblingRepository.findByJournalpostID(journalpostID);
    }

    public Optional<MelosysEessiMelding> finnVedJournalpostIDOpprettMelosysEessiMelding(String journalpostID) {
        Optional<JournalpostSedKobling> journalpostSedKobling = journalpostSedKoblingRepository
                .findByJournalpostID(journalpostID);

        if (journalpostSedKobling.isPresent()) {
            return Optional.of(opprettEessiMelding(journalpostSedKobling.get()));
        }

        Optional<String> rinaSaksnummer = søkEtterRinaSaksnummerForJournalpost(journalpostID);

        if (rinaSaksnummer.isPresent()) {
            return opprettEessiMelding(rinaSaksnummer.get(), journalpostID);
        }

        return Optional.empty();
    }

    public boolean erASedAlleredeBehandlet(String rinaSaksnummer){
        return journalpostSedKoblingRepository.findByRinaSaksnummer(rinaSaksnummer).stream()
            .anyMatch(JournalpostSedKobling::erASed);
    }

    private Optional<String> søkEtterRinaSaksnummerForJournalpost(String journalpostID) {
        Optional<String> rinaSaksnummer = safConsumer.hentRinasakForJournalpost(journalpostID);
        if (rinaSaksnummer.isEmpty()) {
            log.info("Rinasaksnummer er null fra saf for journalpostId: {}", journalpostID);
        }

        return rinaSaksnummer.isPresent() ? rinaSaksnummer : hentFraCaseStore(journalpostID);
    }

    private Optional<String> hentFraCaseStore(String journalpostID) {
        if(unleash.isEnabled(ToggleName.IKKE_HENT_FRA_CASESTORE)) {
            log.info("søkEtterRinaSaksnummerForJournalpost: henting fra casestore er togglet av");
            return Optional.empty();
        }

        return caseStoreConsumer.finnVedJournalpostID(journalpostID)
            .stream().findFirst().map(CaseStoreDto::getRinaSaksnummer);
    }

    private MelosysEessiMelding opprettEessiMelding(JournalpostSedKobling journalpostSedKobling) {
        BUC buc = euxService.hentBuc(journalpostSedKobling.getRinaSaksnummer());
        final var organisation = buc.hentDokument(journalpostSedKobling.getSedId()).getCreator().getOrganisation();
        SED sed = euxService.hentSed(journalpostSedKobling.getRinaSaksnummer(), journalpostSedKobling.getSedId());

        Long gsakSaksnummer = saksrelasjonService.finnVedRinaSaksnummer(journalpostSedKobling.getRinaSaksnummer())
                .map(FagsakRinasakKobling::getGsakSaksnummer)
                .orElse(null);

        return opprettMelosysEessiMelding(
                sed,
                journalpostSedKobling.getSedId(),
                journalpostSedKobling.getRinaSaksnummer(),
                journalpostSedKobling.getSedType(),
                journalpostSedKobling.getBucType(),
                organisation.getId(),
                organisation.getCountryCode(),
                journalpostSedKobling.getJournalpostID(),
                gsakSaksnummer != null ? gsakSaksnummer.toString() : null,
                Integer.parseInt(journalpostSedKobling.getSedVersjon()) != 1,
                journalpostSedKobling.getSedVersjon()
        );
    }

    private Optional<MelosysEessiMelding> opprettEessiMelding(String rinaSaksnummer, String journalpostID) {
        BUC buc = euxService.hentBuc(rinaSaksnummer);
        Optional<Document> documentOptional = buc.hentSistOppdaterteDocument();
        if (documentOptional.isEmpty()) {
            log.warn("Finner ikke sist oppdaterte sed for rinasak {}", rinaSaksnummer);
            return Optional.empty();
        }
        final Document sedDocument = documentOptional.get();
        String sedID = sedDocument.getId();
        String sedType = sedDocument.getType();
        final var organisation = sedDocument.getCreator().getOrganisation();
        SED sed = euxService.hentSed(rinaSaksnummer, sedID);
        final String sedVersjon = "0"; //har ikke sed-versjon

        return Optional.of(opprettMelosysEessiMelding(sed, sedID, rinaSaksnummer, sedType,
                buc.getBucType(), organisation.getId(), organisation.getCountryCode(), journalpostID, null, false, sedVersjon));
    }

    public JournalpostSedKobling lagre(String journalpostID, String rinaSaksnummer, String sedID,
            String sedVersjon, String bucType, String sedType) {
        return journalpostSedKoblingRepository.save(
                new JournalpostSedKobling(journalpostID, rinaSaksnummer, sedID, sedVersjon, bucType, sedType)
        );
    }

    private MelosysEessiMelding opprettMelosysEessiMelding(SED sed, String sedId, String rinaSaksnummer,
                                                           String sedType, String bucType, String avsenderID, String landkode,
                                                           String journalpostID, String saksnummer, boolean erEndring, String sedVersjon) {
        return melosysEessiMeldingMapperFactory.getMapper(SedType.valueOf(sedType))
                .map(null, sed, sedId, rinaSaksnummer, sedType, bucType, avsenderID, landkode, journalpostID,
                        null, saksnummer, erEndring, sedVersjon);
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/journalpostkobling/JournalpostSedKoblingService.kt
package no.nav.melosys.eessi.service.journalpostkobling

import mu.KotlinLogging
import no.nav.melosys.eessi.integration.saf.SafConsumer
import no.nav.melosys.eessi.kafka.producers.model.MelosysEessiMelding
import no.nav.melosys.eessi.models.FagsakRinasakKobling
import no.nav.melosys.eessi.models.JournalpostSedKobling
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.exception.NotFoundException
import no.nav.melosys.eessi.repository.JournalpostSedKoblingRepository
import no.nav.melosys.eessi.service.eux.EuxService
import no.nav.melosys.eessi.service.saksrelasjon.SaksrelasjonService
import no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding.EessiMeldingQuery
import no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding.MelosysEessiMeldingMapperFactory
import org.springframework.stereotype.Service
import java.util.*
import kotlin.jvm.optionals.getOrNull

private val log = KotlinLogging.logger {}

@Service
class JournalpostSedKoblingService(
    private val journalpostSedKoblingRepository: JournalpostSedKoblingRepository,
    private val euxService: EuxService,
    private val saksrelasjonService: SaksrelasjonService,
    private val safConsumer: SafConsumer,
    private val melosysEessiMeldingMapperFactory: MelosysEessiMeldingMapperFactory
) {
    fun finnVedJournalpostID(journalpostID: String): Optional<JournalpostSedKobling> =
        journalpostSedKoblingRepository.findByJournalpostID(journalpostID)

    fun finnVedJournalpostIDOpprettMelosysEessiMelding(journalpostID: String): Optional<MelosysEessiMelding> {
        val journalpostSedKobling = journalpostSedKoblingRepository.findByJournalpostID(journalpostID).getOrNull()
        return Optional.ofNullable(
            journalpostSedKobling?.let { opprettEessiMelding(it) }
                ?: søkEtterRinaSaksnummerForJournalpost(journalpostID)?.let { rinaSaksnummer ->
                    opprettEessiMelding(rinaSaksnummer, journalpostID)
                }
        )
    }

    fun erASedAlleredeBehandlet(rinaSaksnummer: String): Boolean =
        journalpostSedKoblingRepository.findByRinaSaksnummer(rinaSaksnummer).any { it.erASed() }

    fun lagre(
        journalpostID: String,
        rinaSaksnummer: String,
        sedID: String,
        sedVersjon: String,
        bucType: String,
        sedType: String
    ): JournalpostSedKobling =
        journalpostSedKoblingRepository.save(JournalpostSedKobling(journalpostID, rinaSaksnummer, sedID, sedVersjon, bucType, sedType))

    private fun søkEtterRinaSaksnummerForJournalpost(journalpostID: String): String? =
        safConsumer.hentRinasakForJournalpost(journalpostID).getOrNull()?.also {
            log.info("Rinasaksnummer er null fra saf for journalpostId: {}", journalpostID)
        }

    private fun opprettEessiMelding(journalpostSedKobling: JournalpostSedKobling): MelosysEessiMelding {
        val buc = euxService.hentBuc(journalpostSedKobling.rinaSaksnummer)
        val organisation = buc.hentDokument(journalpostSedKobling.sedId).creator?.organisation
            ?: throw IllegalStateException("Organisation er null")
        val sed = hentSed(journalpostSedKobling.rinaSaksnummer, journalpostSedKobling.sedId)
        val gsakSaksnummer = finnVedRinaSaksnummer(journalpostSedKobling)?.gsakSaksnummer
        return opprettMelosysEessiMelding(
            EessiMeldingQuery(
                sed = sed,
                rinaDokumentID = journalpostSedKobling.sedId,
                rinaSaksnummer = journalpostSedKobling.rinaSaksnummer,
                sedType = journalpostSedKobling.sedType,
                bucType = journalpostSedKobling.bucType,
                avsenderID = organisation.id,
                landkode = organisation.countryCode,
                journalpostID = journalpostSedKobling.journalpostID,
                gsakSaksnummer = gsakSaksnummer?.toString(),
                sedErEndring = journalpostSedKobling.sedVersjon.toInt() != 1,
                sedVersjon = journalpostSedKobling.sedVersjon
            )
        ) ?: throw IllegalStateException("Kunne ikke opprette melding for journalpostSedKobling: $journalpostSedKobling")
    }

    private fun finnVedRinaSaksnummer(journalpostSedKobling: JournalpostSedKobling): FagsakRinasakKobling? =
        saksrelasjonService.finnVedRinaSaksnummer(journalpostSedKobling.rinaSaksnummer).getOrNull()

    private fun opprettEessiMelding(rinaSaksnummer: String, journalpostID: String): MelosysEessiMelding? {
        val buc = euxService.hentBuc(rinaSaksnummer)
        val sedDocument = buc.hentSistOppdaterteDocument() ?: run {
            log.warn("Finner ikke sist oppdaterte sed for rinasak {}", rinaSaksnummer)
            return null
        }
        val sedID = sedDocument.id ?: throw IllegalStateException("sedDocument er null")
        val sedType = sedDocument.type
        val organisation = sedDocument.creator?.organisation ?: throw IllegalStateException("Organisation er null")
        val sed = hentSed(rinaSaksnummer, sedID)
        return opprettMelosysEessiMelding(
            EessiMeldingQuery(
                sed = sed,
                rinaDokumentID = sedID,
                rinaSaksnummer = rinaSaksnummer,
                sedType = sedType,
                bucType = buc.bucType,
                avsenderID = organisation.id,
                landkode = organisation.countryCode,
                journalpostID = journalpostID,
                sedErEndring = false,
                sedVersjon = "0" // har ikke sed-versjon
            )
        ) ?: throw IllegalStateException("Kunne ikke opprette melding for rinasak: $rinaSaksnummer")
    }

    private fun hentSed(rinaSaksnummer: String, dokumentId: String) =
        euxService.hentSed(rinaSaksnummer, dokumentId)
            ?: throw NotFoundException("Fant ikke SED med id $dokumentId i rinasak $rinaSaksnummer")

    private fun opprettMelosysEessiMelding(eessiMeldingQuery: EessiMeldingQuery): MelosysEessiMelding? =
        eessiMeldingQuery.sedType?.let { sedType ->
            melosysEessiMeldingMapperFactory.getMapper(SedType.valueOf(sedType)).map(eessiMeldingQuery)
        } ?: log.warn("SedType er null for rinasak: ${eessiMeldingQuery.rinaSaksnummer}").run { null }
}
